### ISSUE TITLE
parallel operator for electronics (infix forbidden)

### ISSUE THREAD

**Issue by Famondir** [maths/moodle-qtype_stack#938]
Hi all,
is there a parallel operator one can use for common calculations for the total resistence in an electric network? I tried to define it myself via
```
infix("||", 120, 120);
"||"(a,b) := a*b/(a+b);
```
but the infix function is forbidden.

We'ld prefer not to define something like `ll(a,b) := a*b/(a+b);` with two small L. It would confuse our students and is pretty cumbersome. And string parsing in JavaScript would be a bit cumbersome for me as the question maintainer to come up with.

Sincerely Simon

**Comment by aharjula** [comment]
Well, this is advanced... Unfortunately, while Maxima allows one to define operators as one wishes, STACK does not currently support this. The reason is simple; I am not yet good enough at coding parsers to deal with arbitrarily defined operators in an efficient way, even less so when those operators contain characters that are used on the supported operators and when those might have different precedence logic.

And we need to have a separate parser outside of Maxima to check certain things and rewrite others before the content actually reaches Maxima. So as long as that parser cannot extend its operator set and deal with new operators with arbitrary precedence, this is not going to happen. However, this is on the wish list as this would be nice and would allow all sorts of fun things, but I doubt we will be getting this in the next parser, maybe the one after it.

Basically, this is not just about `infix` being forbidden. Even if one were to tune the `security-map.json` to allow that, the parser doing our validation would not support it. A suitably adventurous user might then go and tune the `parser-grammar.pegjs` and try to generate a new parser, but that might become difficult fast as the `|` char is already in use for a special syntax-candy feature, and that might make things somewhat hard.

**Comment by Famondir** [comment]
Thanks for your answer. To fill our gap I have adjusted some code of a simple parser in JavaScript that now reads the string in one input field and parses it via onchange event to another hidden input field: [Parsing script](https://videowithstack.4lima.de/htw/Moodle/paralleloperatorparsing.js)

As is it can handle simple math, functions that only use one operator like sin or cos and some example functions like min, max, binomial and ll that need more operators. One can add more functions to it but we found that in tasks where the parallel operator is used other functions are very uncommon.

Pay attention that usage of this script will change all a||b into ll(a,b). Therefore you cant use || as or operator in the same task.

You can find an example question here: [Netzwerkanalyse](https://videowithstack.4lima.de/htw/Moodle/questions/quiz-HA5_1_a-20230502-0941.xml)

**Comment by aharjula** [comment]
Nice. This is indeed a way for doing such things. You can have any syntax you want as long as you are willing to transform it before it becomes STACKs problem.

As a side note, STACK is going towards a full ban on the `<script>`-tag. However, we will still provide means for having scripts that reference inputs and listen and do things with them. Those scripts just need to be declared with a different wrapping and will end up being executed in sandbox iframes. The ban is not coming yet, but you will start seeing the tools that will provide the "safe" alternative this summer, and the ban will follow once those tools have been in play for a while and people have had the chance to migrate their scripts over to that. Your example input handling will not be a problem, but what you do with CSS loading may very well be and may mean that we need to decide whether that is something we wish to allow or not.

**Comment by Famondir** [comment]
At other places I just load the CSS in a Moodle text block. I assume this possibility won't be affected by this ban, would it?

**Comment by aharjula** [comment]
It should be safe. This ban would only affect things inside STACK questions.

Given that it is not uncommon for people to share styles with their materials, it is entirely possible that a mechanism for doing so will be provided. It just cannot happen directly with such free-form JavaScript DOM modification. On the other hand, one can load pretty wild things through CSS documents, so we will need to think this through.

Basically, we want to make it so that no matter where one gets a STACK question and no matter what it contains, it should always be safe to try to use it. At this point, the unsafe bit is JavaScript, which can basically do things as the user, and thus we need to separate it from the possibility to act as the user.
